### DISCUSSION TITLE
Install VS Code Extensions at compile time but not as builtin plugins.

### DISCUSSION THREAD

**Discussion by cj-rajat** [eclipse-theia/theia#11699] [Q&A] [answered]
I have the ids of the vs code extensions that I want to add in Theia before the theia starts. I know how we can add built in plugins using "theiaPlugins" in package.json. As these plugins installed at runtime and can not be uninstalled from UI. Is it possible to download the extensions in the same directory( that I don't know) where theia installs plugins when we install them from the Extension view and when theia starts those plugins are already available to start.
If this is not possible can you suggestt some  way like how I can install plugins i.e how we can fetch the package from the ids of the plugins and install them for the user.

**Comment by msujew** [comment, marked as answer]
Hey @rajattyagi200,

so there would be two ways how you could accomplish this:

1. The default installation path of vscode extensions is at `~/.theia/extensions`
2. When starting your application, you can have a frontend service that does something akin to this:
```ts
@inject(PluginServer)
protected readonly pluginServer: PluginServer;

onStart() {
  const extensionUri = VSCodeExtensionUri.toUri(extension.extension.name, extension.extension.namespace).toString();
  this.pluginServer.deploy(extensionUri);
}
```

See also how [theia installs language packs](https://github.com/eclipse-theia/theia/blob/master/packages/vsx-registry/src/browser/vsx-language-quick-pick-service.ts) for the user during runtime.

**Reply by cj-rajat** [reply]
Thanks for your reply.
I will have a look into this.
One more question is how we can avoid installation if the extension is already available. Is there a method so that we can check if an extension is already available or not? Then only we will download and install the extension.

**Reply by msujew** [reply]
You can ask the `HostedPluginServer` to provide this information for you:

```ts
@inject(HostedPluginServer)
private readonly server: HostedPluginServer;
...
const deployedPluginIds = await this.server.getDeployedPluginIds();
if (!deplyedPluginIds.includes('...') {
    ... 
}
```

**Reply by cj-rajat** [reply]
Thanks
